tests/cupyx_tests/distributed_tests/test_comm_template.py.txt

```diff
@@ -0,0 +1,254 @@
+
+    def run_scatter(rank, n_workers, root, dtype):
+        dev = cuda.Device(rank)
+        dev.use()
+        comm = NCCLBackend(n_workers, rank)
+        in_array = 1 + cupy.arange(
+            n_workers * 10, dtype='f').reshape(n_workers, 10)
+        out_array = cupy.zeros((10,), dtype='f')
+
+        comm.scatter(in_array, out_array, root)
+        if rank > 0:
+            testing.assert_allclose(out_array, in_array[rank])
+
+    _launch_workers(N_WORKERS, run_scatter, (0, dtype))
+    _launch_workers(N_WORKERS, run_scatter, (1, dtype))
+
+
+def gather(dtype):
+    if dtype in (cupy.int16, cupy.uint16):
+        return  # nccl does not support int16
+
+    def run_gather(rank, n_workers, root, dtype):
+        dev = cuda.Device(rank)
+        dev.use()
+        comm = NCCLBackend(n_workers, rank)
+        in_array = (rank + 1) * cupy.arange(10, dtype='f')
+        out_array = cupy.zeros((n_workers, 10), dtype='f')
+        comm.gather(in_array, out_array, root)
+        if rank == root:
+            expected = 1 + cupy.arange(n_workers).reshape(n_workers, 1)
+            expected = expected * cupy.broadcast_to(
+                cupy.arange(10, dtype='f'), (n_workers, 10))
+            testing.assert_allclose(out_array, expected)
+
+    _launch_workers(N_WORKERS, run_gather, (0, dtype))
+    _launch_workers(N_WORKERS, run_gather, (1, dtype))
+
+
+def all_to_all(dtype):
+    if dtype in (cupy.int16, cupy.uint16):
+        return  # nccl does not support int16
+
+    def run_all_to_all(rank, n_workers, dtype):
+        dev = cuda.Device(rank)
+        dev.use()
+        comm = NCCLBackend(n_workers, rank)
+        in_array = cupy.arange(
+            n_workers * 10, dtype='f').reshape(n_workers, 10)
+        out_array = cupy.zeros((n_workers, 10), dtype='f')
+        comm.all_to_all(in_array, out_array)
+        expected = (10 * rank) + cupy.broadcast_to(
+            cupy.arange(10, dtype='f'), (n_workers, 10))
+        testing.assert_allclose(out_array, expected)
+
+    _launch_workers(N_WORKERS, run_all_to_all, (dtype,))
+
+
+def barrier():
+    def run_barrier(rank, n_workers):
+        dev = cuda.Device(rank)
+        dev.use()
+        comm = NCCLBackend(n_workers, rank)
+        comm.barrier()
+        before = time.time()
+        if rank == 0:
+            time.sleep(2)
+        comm.barrier()
+        after = time.time()
+        assert int(after - before) == 2
+
+    _launch_workers(N_WORKERS, run_barrier)
+
+
+def init():
+    def run_init(rank, n_workers):
+        dev = cuda.Device(rank)
+        dev.use()
+        comm = init_process_group(n_workers, rank)
+        # Do a simple call to verify we got a valid comm
+        in_array = cupy.zeros(1)
+        if rank == 0:
+            in_array = in_array + 1
+        comm.broadcast(in_array, 0)
+        testing.assert_allclose(in_array, cupy.ones(1))
+
+    _launch_workers(N_WORKERS, run_init)
+
+
+if __name__ == '__main__':
+    # Run the templatized test
+    {name}({dtype})
```

Comment:
How about using command line arguments instead of templates? I'd like to avoid templates if possible, because it's fragile (e.g. if we write `print(f'{dtype}')` in this file, it break...)


```
func = globals()[sys.argv[1]]
dtype = getattr(cupy, sys.argv[2])
func(dtype)
```

In that case let's name this script `comm_runner.py` or sth (so that flake8 can be applied and pytest does not pick it up as a test script)

Reply:
Done!!